### PR TITLE
Add support for map merge #14913

### PR DESCRIPTION
**Pull request by webashutosh**

### CODE REVIEW

File: hazelcast-sql/src/test/java/com/hazelcast/sql/impl/calcite/OptimizerInstantiationTest.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
+/*
```
Comment (webashutosh) [inline hazelcast-sql/src/test/java/com/hazelcast/sql/impl/calcite/OptimizerInstantiationTest.java:1]: I did not create this class.
I'll remove it from this PR before merge if required

**Comment by webashutosh** [comment]
Hi @mmedenjak - created this PR for supporting the merge() operation. Can you please have a look?
cc @Holmistr 

PS: I believe this will close issue #14913. Do you agree?

**Comment by Holmistr** [comment]
run-lab-run

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1684/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1684/Hazelcast-pr-builder-1684.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] COMPILATION ERROR : 
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast-sql/src/test/java/com/hazelcast/sql/impl/calcite/OptimizerInstantiationTest.java:[53,40] error: getOptimizer() is not public in SqlServiceImpl; cannot be accessed from outside package
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:2.5.1:testCompile (default-testCompile) on project hazelcast-sql: Compilation failure
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast-sql/src/test/java/com/hazelcast/sql/impl/calcite/OptimizerInstantiationTest.java:[53,40] error: getOptimizer() is not public in SqlServiceImpl; cannot be accessed from outside package
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast-sql/src/test/java/com/hazelcast/sql/impl/calcite/OptimizerInstantiationTest.java:[53,40] error: getOptimizer() is not public in SqlServiceImpl; cannot be accessed from outside package
--------------------------
</details>
</pre>

**Comment by Holmistr** [comment]
Hello @webashutosh again :) Thank you very much for another contribution! Let's get the reviews then.

**Comment by webashutosh** [comment]
@Holmistr, thanks! The failing test is not related to my change. My PR has picked `OptimizerInstantiationTest.java` file for some reason (I had pulled and merged upstream changes just prior to raising my PR, so not sure exactly why this happened). I  will check that.

**Comment by Holmistr** [comment]
@webashutosh Just remove it from the PR and repush. Otherwise, we probably won't make a green PR builder.

**Comment by webashutosh** [comment]
@Holmistr done ... thanks for your help as always!

**Comment by Holmistr** [comment]
run-lab-run

**Comment by webashutosh** [comment]
@mmedenjak @Holmistr could you please have a look at this one?

File: hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java [hunk 2]
```diff
@@ -1168,4 +1169,45 @@ private V computeLocally(K key,
...
+        } else {
+            return mergeLocally(key, value, remappingFunction);
+        }
+    }
+
+    private V mergeLocally(K key, V value, BiFunction<? super V, ? super V, ? extends V> remappingFunction) {
+
+        while (true) {
+            Data oldValueAsData = toData(getInternal(key));
+            V oldValueClone = toObject(oldValueAsData);
```
[the remark this answers is not in the document]
Comment (webashutosh) [reply hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java]: Done

File: hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientMapProxy.java [hunk 3]
```diff
@@ -2056,4 +2056,36 @@ private V computeLocally(K key,
...
+        checkNotNull(key, NULL_BIFUNCTION_IS_NOT_ALLOWED);
+
+        return mergeLocally(key, value, remappingFunction);
+    }
+
+    private V mergeLocally(K key, V value, BiFunction<? super V, ? super V, ? extends V> remappingFunction) {
+
+        while (true) {
+            Data oldValueAsData = toData(getInternal(key));
+            V oldValueClone = toObject(oldValueAsData);
```
[the remark this answers is not in the document]
Comment (webashutosh) [reply hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientMapProxy.java]: Done

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/AbstractMapTable.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+package com.hazelcast.sql.impl.schema;
+
+import java.util.List;
+
+/**
+ * A table that is backed by a Hazelcast map.
+ */
+public abstract class AbstractMapTable extends Table {
```
[the remark this answers is not in the document]
Comment (webashutosh) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/schema/AbstractMapTable.java]: Done

File: hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java [hunk 5]
```diff
@@ -107,8 +107,8 @@
 import static com.hazelcast.client.properties.ClientProperty.SHUFFLE_MEMBER_LIST;
 import static com.hazelcast.core.LifecycleEvent.LifecycleState.CLIENT_CHANGED_CLUSTER;
 import static com.hazelcast.internal.nio.IOUtil.closeResource;
-import static com.hazelcast.internal.util.ExceptionUtil.rethrow;
```
[the remark this answers is not in the document]
Comment (webashutosh) [reply hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java]: Done

**Comment by webashutosh** [comment]
> Looks good, added a few minor comments. Sorry for the long delay but if you can take a look at the comments, we'll get it merged this week and include it in the BETA release which is supposed to go out next week. Thank you for all of your effort thus far!

Thanks @mmedenjak ! Pushed a commit with all feedback addressed. Could you please have a look?

**Comment by sancar** [comment]
Hi @webashutosh The commit history looks weird. Can you rebase your changes onto the master so that your commits will not interfere with others? Ideally, we should only see your changes in the commit history.

File: hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java [hunk 6]
```diff
@@ -1168,4 +1169,43 @@ private V computeLocally(K key,
...
+                V newValue = remappingFunction.apply(oldValueClone, value);
+                if (newValue != null) {
+                    if (replaceInternal(key, oldValueAsData, toData(newValue))) {
+                        return newValue;
+                    }
+                } else if (removeInternal(key, oldValueAsData)) {
+                    return null;
+                }
+            } else {
+                Data result =  putIfAbsentInternal(key, toData(value), UNSET, TimeUnit.MILLISECONDS, UNSET,
```
Comment (sancar) [inline hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java:1202]: I think we can avoid deserializing the key twice.
The deserialization happens inside function calls `getInternal` , `replaceInternal `, `putIfAbsentInternal`  and `removeInternal`  .
Instead, we can convert the key once, and pass the same data to all functions. These functions can work with data as the key. 

The same goes for ClientMapProxy. 
The only difference is that you need to use `toDataWithStrategy(key)` on MapProxyImpl for keys.

**Comment by mmedenjak** [comment]
@webashutosh I suggest you use git rebase to rebase your commits on top of latest upstream master. If you want I can try doing it for you on your fork.

**Comment by webashutosh** [comment]
Hi @mmedenjak , no worries, I will do that and push a commit soon.

**Comment by webashutosh** [comment]
Hi @mmedenjak - somehow this PR has been merged and marked closed. I did not do that. Any idea on what might have happenned?

**Comment by webashutosh** [comment]
@mmedenjak  sorry for the confusion. Somehow this PR was closed. I raised a new one with the same changes - https://github.com/hazelcast/hazelcast/pull/17419

**Comment by sancar** [comment]
@webashutosh If I am not mistaken, a pr get closed automatically if there is no difference in the commits with the branch it will be merged to. 
It shows zero commits right now. And the changes are not applied to master yet as far as I see.

**Comment by webashutosh** [comment]
Sure @sancar ... makes sense. I was trying to cleanup the branch. Didn't know that the PR will get auto-closed when diff becomes 0. 
Raised this one - https://github.com/hazelcast/hazelcast/pull/17419
I will tag you on that one once I address your feedback about serializing keys only once

**Comment by webashutosh** [comment]
@sancar @mmedenjak - #17419 is ready for review with all feedback addressed ... thanks!

**Pull request closed without merging**